src/tools/cephfs_mirror/PeerReplayer.cc

```diff
@@ -1197,6 +1197,39 @@ int PeerReplayer::pre_sync_check_and_open_handles(
   return 0;
 }
 
+// Determines if the source is the local (previous) snapshot or the remote dir_root
+int PeerReplayer::validate_source(const std::string &dir_root,
```

Comment:
Sorry - I don't understand why is this required. PeerReplayer would use local snapshot comparison (and hence local mount handles) when it's sure that the data on the remote file system (live) exactly matches the last snapshot transferred. This is determined using an xattr (from the remote file system directory).

Reply:
Right. But this change is made to resolve the test failure https://tracker.ceph.com/issues/69671. This happens only when the source snapshot is deleted during an incremental sync. Then the incremental sync has to check and make the switch to remote dir_root as per `test_cephfs_mirror_sync_with_purged_snapshot`.

Previously, before snapdiff, this was taken care of as everything was a full sync. But with incremental sync, as per the test (test_cephfs_mirror_sync_with_purged_snapshot), the source has to be validated each time. Hope this clarifies.

Reply:
So, the bug was that the mirror daemon was (incorrectly) using incremental sync with remote directory snapshot comparison?

Reply:
@vshankar 
There's no issue with `incremental sync with remote directory snapshot comparison`. But the snapdiff code (the new do_synchronize function) wasn't making the switch to the remote `dir_root`. There was no such mechanism in the new do_synchronize. I think I've missed it back then. That's the reason for this failure https://tracker.ceph.com/issues/69671. With this PR, the switch happens, as per the original design.

Reply:
OK. So, if a snapshot is deleted when it was used for incremental sync, the mirror daemon code was still incorrectly using incremental sync in the next iteration, yes? If that's the case, then can you put this up in the commit message please?

Reply:
Yes it was.
Updated the commit message @vshankar.

Reply:
> @vshankar There's no issue with `incremental sync with remote directory snapshot comparison`. But the snapdiff code (the new do_synchronize function) wasn't making the switch to the remote `dir_root`. There was no such mechanism in the new do_synchronize. I think I've missed it back then. That's the reason for this failure https://tracker.ceph.com/issues/69671. With this PR, the switch happens, as per the original design.

Sorry for the wrong explanation. I didn't miss to implement the switching. This is a regression caused by https://github.com/ceph/ceph/pull/60667/commits/7a747bc55381b840d76202c9897f7a0ab8d143b3. That's the reason `test_cephfs_mirror_sync_with_purged_snapshot` fails now. It was passing earlier, but now it fails.

Reply:
Updated the commit message with the details @vshankar